Changes transfer history/encounter batch loads

fredkingham opened the pull request:
Changes the encounter and transfer history batch loads to work with mergedMRNs.

As part of this I've remove the encounter patient creation from load_excounters_since into update_encounters_from_query_result as otherwise we are doing the demographics lookup twice.

Commit:
Changes transfer history/encounter batch loads

Changes the encounter and transfer history batch loads to work with
mergedMRNs

Commit:
Changes a print statement to a logger statement

Commit:
Adds unit tests for creating/merged transfer histories

Make sure that in the batch we should create patients if they do not
exist.

Commit:
Fixes a typo in the admissions loader test

Commit:
White space fix in the admissions.loader

fredkingham commented:
This still does not deal with merged patients (or at least it does but it will error)

Commit:
Makes us not raise an error is a merged MRN in the upstream result

davidmiller commented on plugins/admissions/loader.py:
```diff
@@ -100,29 +100,26 @@ def cast_to_encounter(encounter, patient):
-        for patient in mrn_to_patients[row['PID_3_MRN'].strip()]:
-            to_create.append(cast_to_encounter(row, patient))
+        mrn = row['PID_3_MRN']
+        if mrn not in mrn_to_patients:
```
Is this how these work? doesn't the way `find_patients_from_mrns` work mean that you have to check again for leading 0s, Nones, empty strings etc?

davidmiller commented on plugins/admissions/loader.py:
```diff
@@ -100,29 +100,26 @@ def cast_to_encounter(encounter, patient):
-        )
-
+    mrn_to_patients = find_patients_from_mrns(mrns)
     to_create = []
```
I would have this variable as `encounters_to_create` given the creation of patients nearby

davidmiller commented on plugins/admissions/loader.py:
```diff
@@ -257,25 +242,23 @@ def create_patients(mrns):
+
+    for mrn in mrns:
+        if mrn not in mrn_to_patients:
+            mrn_to_patients[mrn] = create_rfh_patient_from_hospital_number(
```
Similar.. what if it's e.g. None?

Commit:
Renames a variable in admissions loader

Renames to_create to encounters_to_create so that it doesn't get
confused with any of the patient creation

Commit:
Merge branch 'cerner-merge-branch' into merge-encounters-and-transfer-history

Commit:
Removes upstream transfer history rows with invalid MRNs

We don't want to save upstream rows where the MRN is an empty string,
only 00s or None

Commit:
Remove upstream encounter rows with invalid MRNs

Filter encounter rows with invalid MRNs

Commit:
Fixes a variable name in the admissions tests

Commit:
Merge branch 'cerner-merge-branch' into merge-encounters-and-transfer-history

davidmiller commented on plugins/admissions/loader.py:
```diff
@@ -257,25 +243,35 @@ def create_patients(mrns):
+    for mrn in mrns:
+        if mrn not in mrn_to_patients:
+            mrn_to_patients[mrn] = create_rfh_patient_from_hospital_number(
+                mrn, InfectionService, run_async=False
```
I'm not entirely sure why we're telling this to run_async=False here?

Commit:
Removes run_async=False from the admissions loader

We don't need to create the patients synchronously and none of the other
loaders do.

davidmiller commented on plugins/admissions/loader.py:
```diff
@@ -100,29 +100,27 @@ def cast_to_encounter(encounter, patient):
     existing_encounters = Encounter.objects.filter(upstream_id__in=upstream_ids)
     existing_encounters_count = existing_encounters.count()
     logger.info(f'{existing_encounters_count} existing encounters will be removed')
     mrns = [i['PID_3_MRN'].strip() for i in rows if i['PID_3_MRN'].strip()]
```
What's the reason for this? 
As implemented, find_patients_from does loads of stripping itself, and is_valid_mrn on L103 filters out the need for the if clause?

fredkingham replied:
find_patients_from does do the stripping but the `create_rfh_patient_from_hospital_number` does not so if its a new MRN this would mean that we could still create a patient with MRN 0000

That said, it is superfluous because of line 103.

 It would not be if there were any MRNs in the upstream encounter table with prefixed spaces but there aren't.

davidmiller replied:
My comment was 'no need for strips, no need for python truthiness test' 
We lost one but not the other.
Are you sure the if clause is required? What does it do ?

fredkingham replied:
I think this is resolved so what its doing is clearer, also I think it was sligly confused before. 

We only need to do checks and strips befor creating patients.

Commit:
Changes admissions.loader as we don't need to strip MRNs

The MRNs have had empty MRNs removed by the is_valid_mrn line so we
don't need to do it again.

We don't need to run MRN.strip() as there are no MRNs within the
encounter table with preceding zeros.

Commit:
Changes the admissions loader to be clearer about MRN cleaning

Before creating an MRNs we need to remove Nones and empty strings. The
code prior to this should have been doing this but was not.
